ENG-472 Bridge controller pointer and MintBurnGuard guardrails

piotr-roslaniec opened the pull request:
### Summary
- Governance-set controller pointer on Bridge plus MintBurnGuard to gate controller-driven mint/burn/unmint with caps, pause, and rate limits.
- MintBurnGuard now uses satoshi-denominated accounting with owner-only migration helper and stricter validations to keep exposure in sync with execution targets.

### Motivation
- Provide a least-privilege mint path for integrators while enforcing global guardrails and clear, consistent units across controller flows.

### Changes
- Bridge/BridgeState: controller pointer storage and events; governance-only `setControllerBalanceIncreaser(address)` (zero clears); controller-guarded `controllerIncreaseBalance` and batch variant; expose `controllerBalanceIncreaser()`; storage gap adjusted.
- BridgeGovernance: owner-only forwarder `setControllerBalanceIncreaser(address)` to manage the Bridge controller pointer.
- MintBurnGuard: accounting/storage/events in TBTC satoshis (1e8) with internal 1e18 conversion for Bridge/Bank/Vault calls; removed controller accounting helpers and `reduceExposureAndBurn`; added owner-only `setTotalMinted` (cap-checked, rate window reset); enforced cap vs rate-limit consistency, pause/rate checks, zero-address guards, custom errors, and safe addition (unchecked removed); event naming aligned to `amount`.

piotr-roslaniec committed:
feat: extend bank balance authorization

piotr-roslaniec committed:
docs: add Bridge controller-authorization upgrade explainer; ignore local env; add Sepolia env example

piotr-roslaniec committed:
deploy(upgrade-bridge): prefer cached deployment, use PROXY_ADMIN_PK only, verify library bytecodes before linking

piotr-roslaniec committed:
deploy(transfer-governance): use governance signer (BRIDGE_GOVERNANCE_PK or named account)

piotr-roslaniec committed:
scripts(sepolia): call BridgeGovernance ownership transfer; fix rollback to use governanceDelays(0)

piotr-roslaniec committed:
config(hardhat): set optimizer runs to 200; update comments and keep OZ upgrades plugin + Sepolia env

piotr-roslaniec committed:
deploy(governance): conditional Tenderly/Etherscan verify and minor polish

piotr-roslaniec committed:
chore(docs): format bridge upgrade explainer with Prettier

piotr-roslaniec committed:
chore(solidity): relax lint for deploy/scripts via overrides; fix types; format; make format script run Prettier only to satisfy CI

piotr-roslaniec committed:
Sepolia: redeploy BridgeGovernance and verify allowlist

piotr-roslaniec committed:
Remove unused solidity .env example

piotr-roslaniec committed:
Remove Sepolia-specific upgrade/rollback scripts

piotr-roslaniec committed:
Revert "Sepolia: redeploy BridgeGovernance and verify allowlist"

This reverts commit 766a6d02327d1552767dca4dbb9c2defe8df4de9.

piotr-roslaniec committed:
Remove toggle-increaser helper script

piotr-roslaniec committed:
Remove verify-sepolia helper script

piotr-roslaniec committed:
chore(governance): helper + safer controller sync

piotr-roslaniec committed:
docs: consolidate controller upgrade and governance redeploy

piotr-roslaniec committed:
docs: remove redundant bridge governance redeploy report

piotr-roslaniec committed:
solidity: switch git+ssh deps to git+https

piotr-roslaniec committed:
ci(contracts): use node 18 for solidity jobs

piotr-roslaniec committed:
ci(contracts): align format/slither with node 18

piotr-roslaniec committed:
ci(contracts): move solidity jobs to node 20

piotr-roslaniec committed:
chore(governance): controller ops tooling and docs

piotr-roslaniec committed:
chore(solidity): apply Prettier formatting

piotr-roslaniec committed:
chore(solidity): align hardhat and deploy tooling with main

piotr-roslaniec committed:
chore(solidity): revert unrelated contract formatting

piotr-roslaniec committed:
Fix bridge governance CI and formatting

piotr-roslaniec committed:
Use Node 18 for contracts CI

piotr-roslaniec committed:
chore(governance): harden controller sync and tooling

piotr-roslaniec committed:
ci(docs): ignore engines for solidity docs workflow

piotr-roslaniec committed:
ci(docs): set YARN_IGNORE_ENGINES at workflow level

piotr-roslaniec committed:
ci(contracts): run format job on node 18 and ignore generated docs

jose-blockchain commented on solidity/contracts/bridge/Bridge.sol:
```diff
@@ -1541,6 +1563,72 @@ contract Bridge is
+    }
+
+    /// @notice Batch helper returning authorization status for given addresses.
+    /// @param increasers Addresses to check for controller authorization.
```
empty arrays are cool or not ? just for completeness: maybe add `s.
/// @param increasers Addresses to check. `

jose-blockchain commented on solidity/deploy/80_upgrade_bridge_v2.ts:
```diff
@@ -1,52 +1,123 @@
+  let signerAddress = await signer.getAddress()
+  if (proxyAdminPrivateKey) {
+    signer = new ethers.Wallet(proxyAdminPrivateKey, ethers.provider)
+    signerAddress = await signer.getAddress()
```
maybe log the positive case too:

```typescript
deployments.log(`✓ Using ProxyAdmin signer: ${signerAddress}`)
```

jose-blockchain commented on solidity/test/bridge/Bridge.Governance.test.ts:
```diff
@@ -125,6 +128,150 @@ describe("Bridge - Governance", () => {
+          .to.emit(bank, "BalanceIncreased")
+          .withArgs(recipient(), amount)
+      })
+    })
```
possible diff len test:

```typescript
context("when recipients and amounts arrays have different lengths", () => {
  it("should revert", async () => {
    await bridgeGovernance.connect(governance)
      .setAuthorizedBalanceIncreaser(thirdParty.address, true)
    
    await expect(
      bridge.connect(thirdParty)
        .controllerIncreaseBalances([recipient()], [amount, amount])
    ).to.be.reverted
  })
})
```

jose-blockchain commented on solidity/test/bridge/Bridge.Governance.test.ts:
```diff
@@ -125,6 +128,150 @@ describe("Bridge - Governance", () => {
+      })
+    })
+  })
+
```
maybe some edges cases here: 

```typescript
describe("getAuthorizedBalanceIncreasers", () => {
  it("should handle empty array", async () => {
    const result = await bridge.getAuthorizedBalanceIncreasers([])
    expect(result).to.have.lengthOf(0)
  })
  
  it("should return correct mixed authorization states", async () => {
    // Authorize one, leave another unauthorized, test both
  })
})
```

jose-compu reviewed (changes requested):
This is solid. The core functionality is sound, the upgrade path looks safe, and those mass-revoke safeguards show you're thinking about the operational side. The suggested changes above are all **non-blocking**, they'd make things a bit better but nothing's broken without them.

The main ones I would prioritize:
1. Security comment on `controllerIncreaseBalance` about governance-level power
2. Maybe enforce contract-only controllers (depends on your design intent)
3. Fill in those test gaps for edge cases
4. Enhanced logging in the sync script for better audit trail

LGTM! (moduli comments)

lrsaturnino commented on solidity/contracts/bridge/Bridge.sol:
```diff
@@ -1541,6 +1563,72 @@ contract Bridge is
+    ///      `authorizedBalanceIncreasers` mapping.
+    /// @param recipient Address receiving the balance increase.
+    /// @param amount Amount by which the balance is increased.
+    function controllerIncreaseBalance(address recipient, uint256 amount)
```
I noticed that `controllerIncreaseBalance` only check authorization but don't enforce any amount limits, rate limits, or caps, which means authorized controllers have unrestricted minting power. While the documentation indicates this is intentional and delegates amount validation to controller implementations, I'd like to confirm: is it a requirement that controller contracts implement their own minting limits? If a controller is compromised, the blast radius would be system-wide without Bridge-level circuit breakers. 

I'd consider adding explicit documentation warnings in `IBridgeMintingAuthorization.sol` about this responsibility, or Bridge-level safeguards like per-controller daily mint caps and enhanced monitoring events (Watchdog?).

piotr-roslaniec committed:
feat(bridge): controller caps and MintingGuard scaffold

piotr-roslaniec committed:
refactor(bridge): rely on MintingGuard for caps

piotr-roslaniec committed:
Wire MintingGuard executor and docs for controller allowlist

piotr-roslaniec committed:
test: relax MintingGuard revert assertions

piotr-roslaniec committed:
slither: silence expected reentrancy warnings for controllers

piotr-roslaniec committed:
slither: widen controller reentrancy suppression

piotr-roslaniec committed:
refactor: emit controller events before bank calls

piotr-roslaniec committed:
Wire MintBurnGuard interface, tests, and docs

piotr-roslaniec committed:
chore: rename MintingGuard to MintBurnGuard

piotr-roslaniec committed:
chore: switch bridge to single controller

piotr-roslaniec committed:
fix: share vault interface and satisfy slither

piotr-roslaniec committed:
refactor: keep vault interface local

piotr-roslaniec committed:
Use ITBTCVault in MintBurnGuard and fix lints

piotr-roslaniec committed:
Improve execution wiring and governance transfer workflow

piotr-roslaniec committed:
Allow internal wiring helpers

piotr-roslaniec committed:
refactor: reuse guard increase logic

piotr-roslaniec committed:
Add configurable mint rate limiting to MintBurnGuard

piotr-roslaniec committed:
Suppress time-based lint for MintBurnGuard rate limit

codex committed:
Document MintBurnGuard rate limiting

piotr-roslaniec committed:
fix(mintburn): avoid window start equality

antomor reviewed (commented):
if we plan in future to have contracts with an approach similar to the one adopted by `MintBurnGuard` where we keep track of other contract addresses (e.g.: Bank, Vault, Bridge) maybe we can consider using a Registry to centralize the maintenance and simplify their management. Not sure if it's needed for this though.

piotr-roslaniec committed:
Merge remote-tracking branch 'origin/main' into new/bank-decreaser

# Conflicts:
#	solidity/.prettierignore
#	solidity/hardhat.config.ts

piotr-roslaniec committed:
fix(mintburnguard): address audit findings ENG-473

lrsaturnino reviewed (dismissed):
Took a deep dive into this and it LGTM. The separation between Bridge (who can mint) and MintBurnGuard (caps, pause, rate limits) is clean and makes the security model easy to reason about. I initially flagged a few things as concerns but after actually tracing through the code, they're all intentional - the Bank's handling of zero-address, the batch validation, test coverage - it all checks out. Nice touch on tracking mints AND burns together in totalMinted - that way the cap actually reflects real exposure, not just minting volume.

RobertoMSousa commented on solidity/contracts/test/MockBurnBank.sol:
```diff
@@ -0,0 +1,13 @@
+    uint256 public lastBurnAmount;
+
+    function decreaseBalance(uint256 amount) external {
+        lastBurnAmount = amount;
```
What about adding a new event here to help with debugging the tests?

RobertoMSousa commented on solidity/deploy/44_deploy_mint_burn_guard.ts:
```diff
@@ -0,0 +1,38 @@
+    process.env.MINT_BURN_GUARD_OWNER.length > 0
+      ? process.env.MINT_BURN_GUARD_OWNER
+      : deployer
+
```
Having a validation if the "owner" is a valid address would help stop the process before moving to deployment

miquelcabot committed:
Merge branch 'main' into feat/eng-472-mintburnguard-workstream

miquelcabot committed:
Implement one-off governance function setControllerBalanceIncreaser

miquelcabot committed:
refactor: rename controller-related interfaces, functions and events

miquelcabot committed:
refactor: update MintBurnGuard to use 'operator' terminology instead of 'controller'

miquelcabot committed:
feat: implement MintBurnGuard deployment and configuration scripts

miquelcabot committed:
refactor: add unmintAndBurnFrom and burnFrom functions to MintBurnGuard

miquelcabot committed:
refactor: enhance unmintAndBurnFrom and burn functions with SafeERC20 usage

miquelcabot committed:
refactor: add error handling for bank transfer failure in MintBurnGuard

miquelcabot committed:
test: add operator change functionality tests in MintBurnGuard

miquelcabot committed:
refactor: rename controller to operator in MintBurnGuard tests

miquelcabot committed:
feat: implement MockAccountControl for testing MintBurnGuard integration

miquelcabot committed:
feat: add bank and tbtcToken addresses for improved vault integration, and modify IBank.transferBalanceFrom() signature,

miquelcabot committed:
refactor: change TBTC token approval to use safeIncreaseAllowance in MintBurnGuard

miquelcabot committed:
fix: update balance handling in MintBurnGuard to use satoshi (1e8) amounts instead of tbtc (1e18)

miquelcabot committed:
Merge branch 'main' into new/bank-decreaser

miquelcabot committed:
Merge branch 'new/bank-decreaser' into feat/eng-472-mintburnguard-workstream

miquelcabot committed:
feat: add bridge address to MintBurnGuard for improved vault integration

miquelcabot committed:
feat: refactor MintBurnGuard to use upgradeable proxy pattern and add state management library

miquelcabot committed:
feat: update Slither version to 0.10.1

miquelcabot committed:
feat: update Slither version to 0.11.0

miquelcabot committed:
feat: add 'incorrect-equality' to Slither detectors to exclude

miquelcabot committed:
feat: update Slither command to fail on high severity issues

miquelcabot committed:
feat: add integration tests for MintBurnGuard contract interactions

piotr-roslaniec committed:
ENG-472: MintBurnGuard workstream. Governance, interfaces, integration & tests (#921)

## Summary
This PR completes the ENG-472 workstream by implementing the
`MintBurnGuard` contract with full integration support for Account
Control, comprehensive testing, deployment scripts, and Bridge
governance wiring.

The focus is to establish a robust controller-driven mint/burn path with
global exposure tracking, rate limiting, and pause controls while
maintaining clean separation between the guard's enforcement logic and
Account Control's reserve management.

## Scope
- **Bridge controller authorization**: Implemented
`setMintingController()` governance function and renamed
`IMintingAuthorization` --> `IBridgeController` for clarity.
- **Operator terminology standardization**: Unified naming from
"controller" to "operator" throughout MintBurnGuard to distinguish from
Bridge's "minting controller" role.
- **Core burn/unmint functionality**: Implemented `unmintAndBurnFrom()`
and `burnFrom()` methods for AccountControl integration, with SafeERC20
usage and comprehensive error handling.
- **Interface refinement**: Updated `IMintBurnGuard` with new
burn/unmint methods; extended `IBank` and `ITBTCVault` interfaces for
integration requirements.
- **Deployment infrastructure**: Created deployment and configuration
scripts (`44_deploy_mint_burn_guard.ts`,
`45_configure_mint_burn_guard.ts`,
`46_authorize_mint_burn_guard_in_bridge.ts`).
- **Comprehensive test coverage**: Expanded from 28 to 38 tests,
including 10 new integration tests with `MockAccountControl` simulating
realistic AccountControl usage patterns.
- **Mock contracts for testing**: Created `MockAccountControl`,
`MockBridgeController`, `MockBurnBank`, and `MockBurnVault` to support
integration testing.

## Motivation
ENG-472 introduced a least-privilege operator pattern with global
mint/burn guardrails. This PR builds on that foundation by:
- Completing the MintBurnGuard implementation with all required
burn/unmint methods
- Establishing clear terminology (operator vs. controller) to avoid
confusion
- Providing comprehensive integration testing that validates
AccountControl usage patterns
- Creating production-ready deployment scripts with governance
authorization flow
- Ensuring the design is aligned with Bridge governance and Account
Control expectations

The completed implementation enables AccountControl to manage
per-reserve accounting while MintBurnGuard enforces system-wide caps,
rate limits, and pause controls.

## Notes
- This PR is the completion of the initial ENG-472 MintBurnGuard
workstream.
- **Breaking changes**: Interface renames (`IMintingAuthorization` -->
`IBridgeController`, `controller` --> `operator` in MintBurnGuard)
require migration for any external integrations.
- **Test coverage**: All 38 tests passing with comprehensive coverage of
mint/burn flows, operator authorization, cap enforcement, rate limiting,
and multi-reserve scenarios.

antomor committed:
fix(bridge): re-upgrade Bridge proxy to MintBurnGuard-aware   implementation (#936)

# Summary

PR#933 upgraded the Sepolia Bridge proxy to repair rebateStaking (using
reinitializer(5)), overwriting a prior deployment from
new/bank-decreaser that
had introduced MintBurnGuard controller support. The current on-chain
Bridge is
  missing:

  - setMintingController() / getMintingController()
  - controllerIncreaseBalance() / controllerIncreaseBalances()

  This PR adds the tooling to re-upgrade the Sepolia Bridge proxy to the
new/bank-decreaser implementation, restoring those methods. No
reinitializer is
  called — the mintingController slot defaults to address(0) and is set
  post-upgrade via governance.

##  Changes

- solidity/deploy/84_upgrade_bridge_mint_burn_controller.ts — new deploy
script
that reads the ProxyAdmin from the EIP-1967 admin slot, deploys a fresh
Bridge
  implementation linked to the existing Sepolia libraries, and calls
ProxyAdmin.upgrade() directly. Avoids dependency on the OZ network
manifest.
- solidity/deploy/utils/library-resolution.ts — fix
verifyLibraryBytecodes to
use FQN contracts/bridge/${name}.sol:${name} to avoid HH701 ambiguity on
the
  Wallets artifact.
- solidity/test/bridge/Bridge.MintBurnControllerUpgrade.test.ts — fork
test
validating the upgrade against a Sepolia fork. Requires Anvil (--network
system_tests); also activates when HARDHAT_NETWORK=system_tests so
FORKING_URL
  is not required.
- solidity/scripts/snapshot.sh — pre-upgrade snapshot script: records
baseline
  on-chain state and validates abort conditions.
- solidity/scripts/verify-upgrade.sh — post-upgrade verification script:
asserts
implementation changed, new methods exist, rebateStaking preserved,
governance
  and ProxyAdmin unchanged.
- solidity/.env.example — documents all env vars required for the
upgrade
  workflow.
- docs/bridge-mintburn-controller-upgrade-runbook.md — end-to-end
8-phase
runbook covering snapshot, build, standard tests, fork test, dry run
with real
  keys, live upgrade, post-upgrade verification, and skip guard restore.

##  Runbook status

  - [x] Phase 1 — Pre-upgrade snapshot (snapshot.sh)
  - [x] Phase 2 — Build and lint
  - [x] Phase 3 — Standard test suite
  - [x] Phase 4 — Fork test (4/4 passing against Sepolia fork on Anvil)
- [x] Phase 5 — Dry run with real keys against Anvil fork (--network
system_tests)
  - [x] **Phase 6 — Execute the live upgrade on Sepolia**
  - [x] Phase 7 — Post-upgrade verification (verify-upgrade.sh)
  - [x] Phase 8 — Re-enable func.skip and commit

piotr-roslaniec committed:
fix(deploy): TOB-TBTCAC-45: enforce strict library bytecode verification in Bridge upgrade

verifyLibraryBytecodes() defaulted to non-strict mode, logging warnings
on bytecode mismatches instead of aborting. This allowed Bridge proxy
upgrades to proceed with unreviewed or modified external libraries
(Deposit, DepositSweep, Redemption, Wallets, Fraud, MovingFunds).

Change the default to strict=true and pass strict explicitly in the
upgrade script so any mismatch aborts before upgradeProxy executes.

piotr-roslaniec committed:
fix(deploy): TOB-TBTCAC-45: enforce strict library bytecode verification (#950)

## Summary

- Changes `verifyLibraryBytecodes()` default from `strict=false` to
`strict=true` so bytecode mismatches abort instead of logging warnings
- Passes `strict=true` explicitly in `80_upgrade_bridge_v2.ts` to make
intent clear
- Prevents Bridge proxy upgrades from proceeding with unreviewed or
modified external libraries (Deposit, DepositSweep, Redemption, Wallets,
Fraud, MovingFunds)

## Context

Trail of Bits finding **TOB-TBTCAC-45**: "Non-fatal library verification
allows Bridge upgrades to link unreviewed external libraries". The
`verifyLibraryBytecodes` utility logged warnings on bytecode mismatches
but did not abort, allowing `upgradeProxy` to execute with potentially
compromised library contracts.

## Test plan

- [ ] Verify upgrade script aborts when a library bytecode mismatch is
detected (mock a stale library deployment)
- [ ] Verify upgrade script succeeds when all library bytecodes match
- [ ] Verify existing tests pass with no regressions

piotr-roslaniec committed:
chore: merge main into new/bank-decreaser

Brings in rebate staking, cross-chain updates, and TypeScript SDK
changes from main while preserving the MintBurnGuard / bank-decreaser
feature commit history.
